README.md
## NODEJS API

> A starting point to implement APIs with NodeJS and mongoDB.


## Table of Contents

  1. [Prerequisites](#prerequisites)
  1. [Install dependencies](#install-dependencies)
  1. [Workflow](#workflow)
  1. [Style Guides](#style-guides)
  1. [Contributing](#contributing)
  1. [Changelog](#changelog)
  1. [License](#license)


## Prerequisites

- [Node.js](https://nodejs.org/en/download/).
- [npm](https://www.npmjs.com/), installed with Node.js.
- [mongoDB](https://www.mongodb.com/download-center?jmp=nav#community).


## Install dependencies

#### 1) Check `Node.js` version.

```sh
node --version
```
The version should be at or above 6.10.3

If you don't have Node.js installed go to [nodejs](https://nodejs.org/en/download/) and install the appropiate version or use [nvm](http://www.sergiolepore.net/2014/06/30/nvm-instalando-y-usando-node-version-manager/) (Recommended).

#### 2) Check `npm` version.

```sh
npm --version
```
The version should be at or above 5.0.0

Update npm version

```sh
npm install -g npm
```

#### 3) Install `mongoDB`
The version should be at or above 3.


## Workflow

  1. [Environment variables](#environment-variables)
  1. [Development workflow](#development-workflow)
  1. [Production workflow](#production-workflow)
  1. [Handle dependencies](#handle-dependencies)
  1. [Scripts](#others-scripts)

### Environment variables

```sh
##
# nodejs-api environment variables
##

# Host configuration
## Environment for server.
export NODE_ENV="development"
## Port for api.
export NODEJS_API_API_PORT=8080

## A list with the name of service (folder) to load. Separate by space.
export NODEJS_API_LOADER=undefined

# Database configuration if use a unique db for all services.
## Database URI.
export NODEJS_API_MONGODB_URI="mongodb://localhost:27017/nodejs_api"
```

You can copy this script in bashrc file, modify the variables and delete variables that don't need modification or are undefined.

To change any of this variables:

```sh
# Select services to load
export NODEJS_API_LOADER="folder1 folder2 folder3"

# Select only routes service
export NODEJS_API_LOADER="routes"
```

### Environment variables by sevice

#### Services
- [user](services/user/README.md)

### Development workflow

#### Install the workspace dependencies.

```sh
# cd to project folder
npm install
```

#### Start with npm script

```sh
# cd to project folder
npm start
```

#### Use nodemon tool

```sh
# install nodemon globally
npm install nodemon -g
# cd to project folder
nodemon
```

#### Document with RAML

Write documentation in `raml` folder then run:

```sh
# cd to project folder
npm run document
```

The script create and html documentation in `services/routes/views/docs/docs.html` path.

### Production workflow
- Set the environment variables with production values. `export NODE_ENV='production'`, ...
- Install workspace dependencies.

```sh
# cd to project folder
npm install

# Or if need install development dependencies run
npm run install:dev
```

#### Use PM2 process manager.

```sh
# cd to project folder
pm2 start ecosystem.config.js
```

> If update the repository repeat the process since `Install workspace dependencies` step.

### Handle dependencies
Follow this rules to update dependencies:

- Install dependencies

```sh
# To install production dependencies
# cd to project folder
npm install express

# To install development dependencies
# cd to project folder
npm install eslint -D
```

- Uninstall dependencies

```sh
# cd to project folder
npm uninstall express
```

- Check outdated dependencies

```sh
# cd to project folder
npm outdated --long
```

- **To Upgrade any dependencies follow `Uninstall dependencies` step and install again like `Install dependencies`.**

### Others scripts

```sh
# Run eslint
npm run lint

# Run all tests
npm test

# Run api tests
npm run test:api

# Run unit tests
npm run test:unit
```


## Style Guides

### JavaScript

**[Nimedev JavaScript style guide](https://github.com/nimedev/javascript)**.


## [Contributing](CONTRIBUTING.md)


## [Changelog](CHANGELOG.md)


## [License](LICENSE.md)


CHANGELOG.md
# Change Log

## [10.0.0] - 2017-08-22

### Modified
- Move test from __test__ to e2e folder.

### Updated
— Update package.json engines.
— Upgrade devDep: eslint eslint-config-nimedev-base.


## [9.0.0] - 2017-08-14

### Modified
- Use a unique database as default configuration.

### Updated
- Upgrade dep: mongoose.


## [8.3.4] - 2017-08-12

### Updated
- Upgrade dep: express mongoose.
- Upgrade devDep: eslint.
- Update package engines.


## [8.3.3] - 2017-08-06

### Updated
- Upgrade dep: mongoose.
- Upgrade devDep: chai eslint mocha raml2html.


## [8.3.2] - 2017-07-29

### Updated
- Update engines versions.


## [8.3.1] - 2017-07-23

### Updated
- Upgrade dep: compression cors mongoose.
- Upgrade devDep: chai eslint eslint-plugin-import raml2html.


## [8.3.0] - 2017-06-25

### Deleted
- Remove pre-commit hook.

### Modified
- Import some modules usign destructuring.

### Updated
- Upgrade to eslint v4.
- Upgrade dep: mongoose.


## [8.2.1] - 2017-06-11

### Updated
- Upgrade dep: mongoose.
- Upgrade devDep: chai.


## [8.2.0] - 2017-06-04

### Added
- Implement user.connection module to create the db connection for this service.


## [8.1.7] - 2017-06-02

### Fixed
- Fix install:dev script.
- Fix links in workflow section.

### Modified
- Change name of user service database.

### Updated
- Upgrade dep: mongoose.
- Upgrade devDep: chai eslint-config-nimedev-base.


## [8.1.6] - 2017-05-29

### Fixed
- Fix README.md.


## [8.1.5] - 2017-05-29

### Fixed
- Fix enumeration in install dependencies steps.


## [8.1.4] - 2017-05-29

### Modified
- Upgrade to npm v5.0.0 and update README.md documentation.

### Updated
- Upgrade dep: body-parser express mongoose.
- Upgrade devDep: eslint-config-nimedev-base eslint-plugin-import mocha morgan.


## [8.1.3] - 2017-05-23

### Fixed
- Fix some comments.


## [8.1.2] - 2017-05-21

### Fixed
- Fix name of user.schema file.


## [8.1.1] - 2017-05-21

### Modified
- Hide console.log in server when NODE_ENV is equal to 'test'.


## [8.1.0] - 2017-05-21

### Modified
- Rename user.tools to user.dam.


## [8.0.1] - 2017-05-21

### Added
- Add NODEJS_API_API_PORT to README file.


## [8.0.0] - 2017-05-20

### Deleted
- [Breaking change] Delete NODEJS_API_LAMBDA environment variable.

### Modified
- [Breaking change] Rename functions folder to services.


## [7.2.0] - 2017-05-19

### Added
- Implement detection of services in funcitons folder.

### Modified
- Rename the environment variable used to load services.


## [7.1.0] - 2017-05-19

### Added
- Implement dynamic loading of services.
- Add RAML documentation.


## [7.0.0] - 2017-05-19

### Modified
- [Breaking change]: rebuild the file structure to use with AWS lambda.


## [6.0.0] - 2017-05-19

### Added
- Add ecosystem.config.js file to run PM2

### Modified
- [Breaking change]: change to a file structure as a service.


## [5.0.0] - 2017-05-08

### Modified
- Update database module.
- Update error-handler module.
- Breaking Change: Rename components folder to services.
- Move controller test to main test folder.

### Updated
- Upgrade dep: cors mongoose.
- Upgrade devDep: eslint eslint-config-nimedev-base mocha.


## [4.13.1] - 2017-03-29

### Updated
- Upgrade dep: body-parser cors express mongoose.
- Upgrade devDep: eslint.


## [4.13.0] - 2017-02-27

### Updated
- Upgrade dep: body-parser and mongoose.
- Upgrade devDep: eslint and morgan.


## [4.12.0] - 2017-02-10

### Updated
- Upgrade devDep: eslint and morgan.


## [4.11.0] - 2017-02-03

### Deleted
- Remove .jsbeautifyrc file.

### Updated
- Upgrade devDep: eslint.
- Update dep: express and mongoose.
- Updated to nodejs 6.9.5.


## [4.10.0] - 2017-01-24

### Added
- Add .jsbeautifyrc file to control vscode autoformating.

### Updated
- Upgrade mongoose and eslint.


## [4.9.0] - 2017-01-19

### Modified
- Implement tests for user.dao module.

### Updated
- Upgrade dep: body-parser.
- Upgrade to yarn 0.19.1.


## [4.8.1] - 2017-01-17

### Modified
- Rename user.schema file to user-schema.


## [4.8.0] - 2017-01-16

### Added
- Implement errorLogger function.

### Updated
- Upgrade devDep: eslint-config-nimedev-base.
- Upgrade dep: mongoose.


## [4.7.1] - 2017-01-15

### Fixed
- Improve documentation in workflow section.

### Modified
- Update vscode settings.


## [4.7.0] - 2017-01-14

### Modified
- Improve workflow documentation.


## [4.6.0] - 2017-01-13

### Fixed
- Fix bug in with paging in database-service.

### Modified
- Improve error handling.
- Move morgan and errorhandler to devDependencies.


## [4.5.0] - 2017-01-12

### Modified
- Update database-service module.

### Updated
- Upgrade eslint-config-nimedev-base.


## [4.4.0] - 2017-01-12

### Added
- Add link to javascript style guide in README.md file.

### Modified
- Rebase database module.
- Rename config-express modules to express-setup.
- Reorganize config module.


## [4.3.0] - 2017-01-10

### Modified
- Improve documentation for environment variables.

### Updated
- Upgrade eslint.


## [4.2.0] - 2017-01-05

### Updated
- Upgrade mongoose and eslint-config-nimedev-base.
- Upgrade NodeJS to v6.9.4.


## [4.1.0] - 2016-12-28

### Added
- Add .nvmrc file.


## [4.0.0] - 2016-12-28

### Modified
- Implement eslint-config-nimedev-base.
- Add .editorconfig file.

### Deleted
- Remove unnecesary vscode settings.
- Remove unnecesary 'use strict'.


## [3.2.1] - 2016-12-27

### Updated
- Upgrade mongoose.


## [3.2.0] - 2016-12-20

### Added
- Implement eslint-config-nimedev-node.

### Updated
- Upgrade mongoose, eslint and pre-commit.


## [3.1.1] - 2016-12-12

### Updated
- Upgrade mongoose and eslint.
- Use node 6.9.2.


## [3.1.0] - 2016-12-05

### Updated
- Upgrade mongoose, eslint and mocha.


## [3.0.0] - 2016-11-20

### Modified
- Rebase database module and put database configuration in database module.


## [2.4.0] - 2016-11-16

### Modified
- Improve user tests.

### Updated
- Upgrade eslint, mongoose and errorhandler.


## [2.3.0] - 2016-11-09

### Deleted

config.js
/**
 * @module server.config
 */

'use strict'

const fs = require('fs')
const path = require('path')

const { normalizeUri } = require('./libs/database-tools')

const getDirectories = srcpath => (
  fs.readdirSync(srcpath)
    .filter(file => fs.lstatSync(path.join(srcpath, file)).isDirectory())
)

// A list with the folder name of the services to load
const defaultServices = getDirectories(path.join(__dirname, 'services'))

const readServices = () => {
  const services = process.env.NODEJS_API_LOADER
  return services ? services.split(' ') : defaultServices
}

const config = {
  port: process.env.OPENSHIFT_NODEJS_PORT || process.env.NODEJS_API_PORT || 8080,
  ip: process.env.OPENSHIFT_NODEJS_IP || undefined,
  services: readServices(),

  // Database configuration if use a unique db for all services.
  defaultMongoUri: normalizeUri(process.env.NODEJS_API_MONGODB_URI) || normalizeUri('mongodb://localhost:27017/nodejs_api'),
}

// Test and log app settings
if (process.env.NODE_ENV !== 'test') {
  // eslint-disable-next-line no-console
  console.log('\nserver.config:', JSON.stringify(config, null, ' '))
}

module.exports = Object.freeze(config)